apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1262 @@
+
+    bool passing = false;
+    int passes = 0;
+
+    while (passing == false) // Multiple passes are needed when steepness problems arise for both x and y axis simultaneously
+    {
+        passing = true;
+        for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+        {
+            limitAlteredHeights(cellCoordinates);
+        }
+        std::reverse(mAlteredCells.begin(), mAlteredCells.end()); //Instead of alphabetical order, this should be fixed to sort cells by cell coordinates
+        for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+        {
+            if (!limitAlteredHeights(cellCoordinates, true)) passing = false;
+        }
+        ++passes;
+        if (passes > 2)
+        {
+            Log(Debug::Warning) << "Warning: User edit exceeds accepted slope steepness. Automatic limiting has failed, edit has been discarded.";
+            if (CSVRender::PagedWorldspaceWidget *paged =
+                dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+            {
+                paged->resetAllAlteredHeights();
+                mAlteredCells.clear();
+                return;
+            }
+        }
+    }
+
+    undoStack.beginMacro ("Edit shape and normal records");
+
+    for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+    {
+        std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY());
+        undoStack.push (new CSMWorld::TouchLandCommand(landTable, ltexTable, cellId));
+        const CSMWorld::LandHeightsColumn::DataType landShapePointer = landTable.data(landTable.getModelIndex(cellId, landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandMapLodColumn::DataType landMapLodPointer = landTable.data(landTable.getModelIndex(cellId, landMapLodColumn)).value<CSMWorld::LandMapLodColumn::DataType>();
+        CSMWorld::LandHeightsColumn::DataType landShapeNew(landShapePointer);
+        CSMWorld::LandMapLodColumn::DataType mapLodShapeNew(landMapLodPointer);
+        for(int i = 0; i < ESM::Land::LAND_SIZE; ++i)
+        {
+            for(int j = 0; j < ESM::Land::LAND_SIZE; ++j)
+            {
+                if (CSVRender::PagedWorldspaceWidget *paged =
+                    dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+                {
+                    if (paged->getCellAlteredHeight(cellCoordinates, i, j))
+                        landShapeNew[j * ESM::Land::LAND_SIZE + i] = landShapePointer[j * ESM::Land::LAND_SIZE + i] + *paged->getCellAlteredHeight(cellCoordinates, i, j);
+                    else
+                        landShapeNew[j * ESM::Land::LAND_SIZE + i] = 0;
+                }
+            }
+        }
+        for(int i = 0; i < ESM::Land::LAND_GLOBAL_MAP_LOD_SIZE; ++i)
+        {
+            mapLodShapeNew[i] = landMapLodPointer[i]; //TO-DO: generate a new mWnam based on new heights
+        }
```

Comment:
```cpp
std::copy(landMapLodPointer, landMapLodPointer + ESM::Land::LAND_GLOBAL_MAP_LOD_SIZE, mapLodShapeNew);
```

Reply:
I gave the idea thumbs up, but this very particular `for` needs to be developed to generate a new mWnam record. I'd leave it as it is, until I figure out a proper way for lod map generation.